src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java

```diff
@@ -308,6 +310,33 @@ public void onSetSize(float width, float height, CallbackInfo ci) {
         }
     }
 
+    @Inject(method = "setDead", at = @At("HEAD"))
+    public void onSetDead(CallbackInfo ci) {
+        net.minecraft.entity.Entity mcEntity = (net.minecraft.entity.Entity) (Object) this;
+        if (!this.world.isRemote && (!(mcEntity instanceof EntityLivingBase) || (mcEntity instanceof EntityArmorStand))) {
+            IMixinWorldServer spongeWorld = (IMixinWorldServer) this.world;
+            final CauseTracker causeTracker = spongeWorld.getCauseTracker();
+
+            PhaseData phaseData = causeTracker.getCurrentPhaseData();
+
+            Cause.Builder causeBuilder = Cause.builder();
+            Optional<Object> sourceOptional = phaseData.context.getSource(Object.class);
+            sourceOptional.ifPresent(source -> {
+                causeBuilder.named(NamedCause.source(source));
+                if (source instanceof Projectile) {
+                    ProjectileSource projectileSource = ((Projectile) source).getShooter();
+                    causeBuilder.named(NamedCause.of(NamedCause.THROWER, projectileSource));
+                }
+            });
+            Optional<ItemStack> usedItemOptional = phaseData.context.firstNamed("ItemUsed", ItemStack.class);
+            usedItemOptional.ifPresent(item -> causeBuilder.named(NamedCause.of("ItemUsed", item)));
```

Comment:
Which part of it do you want multiline? Nothing there should really ever throw an exception.